### DISCUSSION TITLE
How do I debug gradient calculation?

### DISCUSSION THREAD

**Discussion by NeilGirdhar** [jax-ml/jax#23046] [Q&A]
I'm having a strange issue whereby two functions that are nearly identical and produce identical primal outputs give extremely different gradients.
```python
import jax.numpy as jnp
from jax import tree, value_and_grad
from jax.experimental import enable_x64
from tjax import JaxRealArray, print_generic

from efax import GammaEP, GammaVP


def test_gamma_vp_entropy_gradient() -> None:
    def f(x_e: GammaEP) -> JaxRealArray:
        x_n = x_e.to_nat()
        return x_e.cross_entropy(x_n)

    def g(x_e: GammaEP) -> JaxRealArray:
        x_n = x_e.to_nat()
        x_e2 = x_n.to_exp()
        print_generic(x_e=x_e, x_e2=x_e2)  # Same!
        return x_e2.cross_entropy(x_n)

    o = jnp.ones(())
    x = GammaVP(o, o * 0.25)
    x_e = x.to_exp()
    f_value, f_gradient = value_and_grad(f)(x_e)
    g_value, g_gradient = value_and_grad(g)(x_e)
    assert tree.all(tree.map(jnp.allclose, f_value, g_value))  # Same primals!
    print_generic(f=f_gradient, g=g_gradient)
    assert tree.all(tree.map(jnp.allclose, f_gradient, g_gradient))  # Fails!


with enable_x64():
    test_gamma_vp_entropy_gradient()
```

It seems really odd that adding a round-trip conversion from `x_e` to `x_n` and back ruins the gradients, but leaves the primals unchanged.

This conversion relies on a simple use of a fixed point optimizer.  I tried substituting jaxopt and optimistix, but they had very poor results.
